Issue #3954: Spooled Window Sinks

hawkfish opened the pull request:
Convert the `Sink` stage of `PhysicalWindow` to use buffered storage instead of in-memory `ChunkCollection`s.

There are two performance regressions in TPC-DS (Q47 and Q57) caused by memory overflow. Since only 7% of the pages are allocated by window, it is arguable that hash joins are the real culprit. Anther fix would be to find a way to increase the memory allocation for the CI task (it seems to be 8GB at the moment).

Commit:
Issue #3453: Early Window Partitions

Move partitioning to the Sink phase so we only have to scan for it once.
Partition into global sorts, to remove an unnecessary copy.
Checkpoint commit of compiling but untested code.

Commit:
Issue #3453: Early Window Partitions

Move partitioning to the Sink phase so we only have to scan for it once.
Partition into global sorts, to remove an unnecessary copy.
Checkpoint commit of compiling but untested code.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Issue #3453: Early Window Partitions

Fix some of the failures, starting with building the partition selections.
Checkpoint commit of partially tested code.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Issue #3453: Early Window Partitions

Checkpoint commit.
Fix all the crash failures.
Fix some sorting.
Still some incorrect results at scale.

Commit:
Merge from master.

Commit:
Issue #3453: Early Window Partitions

Fix Combine and GetData so they don't lose data.
All tests now pass.

Commit:
Merge branch 'hawkfish-small-partition' of github.com:hawkfish/duckdb into hawkfish-small-partition

Commit:
Issue #3453: Early Window Partitions

Rework hash groups into separate objects.
Only use thread-local local sorts for large groups.
Perf regression for window_partition.benchmark down to 1.8x,
gated by Sink thread count.

Commit:
Issue #3453: Early Window Partitions

Release memory more aggressively.

Commit:
Issue #3453: Early Window Partitions

Expand buffer pool allocation error message.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Issue #3954: Constant hash vectors

Correctly handle a constant hash vector.
Another victory for vector size 2.

Commit:
Issue #3954: TPCH_DS Q47 Memory

Allow this benchmark to spill to disk as the new Window buffering pushed it over the edge.
Almost all the query pages (86%) are used by the hash aggregation,
but Window used to just use regular memory, which did not trigger OOM.

Commit:
Issue #3954: TPCH-DS Benchmark Memory

Allow storage for all benchmarks now as there are other failures.

Commit:
Merge branch 'master' into hawkfish-small-partition

Mytherin commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -21,63 +22,434 @@ namespace duckdb {
+		// We take 1/4th of this, to be conservative
+		idx_t max_memory = buffer_manager.GetMaxMemory();
+		idx_t num_threads = TaskScheduler::GetScheduler(context).NumberOfThreads();
+		memory_per_thread = (max_memory / num_threads) / 4;
```
This seems to be copy pasted from `physical_order.cpp`, perhaps turn this into a method there? Especially since it is a slightly evil heuristic it would be good to have it in one place :)

hawkfish replied:
It's a bit worse than that... 

Since each thread can buffer up to 1024 sets of data, the local sorts should really be flushed with 1 / 1024 of that much memory. But the original heuristic should be shared.

Mytherin reviewed (commented):
Thanks for the PR! The code looks good to me minus one small nitpick.

Could we add a few tests that test this behavior? For example, we could set a low memory limit and run a window on a bigger-than-that-memory-limit dataset. Or perhaps the same strategy as the ORDER BY could be used using `debug_force_external`.

Perhaps @lnkuiper also wants to have a look when he is back from holiday (next week?).

Commit:
Issue #3954: Sorting Tasks

Add sorting tasks to move sorting out of the Source phase
and prevent it from being single threaded.

Commit:
Issue #3954: Fix Combine segfault

Check that the local sink state actually has rows before trying to merge them.

hawkfish commented:
> Could we add a few tests that test this behavior? For example, we could set a low memory limit and run a window on a bigger-than-that-memory-limit dataset. Or perhaps the same strategy as the ORDER BY could be used using `debug_force_external`.

It looks like `test_window_sort.test_slow` already does this?

Commit:
Issue #3954: External Window Sorting

Add support for the force_external_sorting PRAGMA.
Share the current heuristic for thread memory.

Commit:
Issue #3954: Tidy fix

Somehow this missing LF snuck into master...

Commit:
Merge branch 'hawkfish-small-partition' into hawkfish-window-sorting

lnkuiper reviewed (approved):
LGTM!

Richard is correct that `test_window_sort.test` tests external sorting for the window operator, as the `external` flag of the sort state in the global window state is set using the `force_external` flag from the client config.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Merge branch 'hawkfish-small-partition' into hawkfish-window-sorting

Commit:
Issue #3954: External Window Sorting

Don't ConcatenateBlocks single blocks - copy instead.
Don't copy order data when sort copies it again.

Commit:
Issue #3954: External Window Sorting

Parallel scan infrastructure for PhysicalHashAggregate.

Commit:
Issue #3954: External Window Sorting

Move mutex to where it is actually needed.

Commit:
Issue #3954: External Window Sorting

Remove LocalSortStates
Track stacked window thread count.
Code cleanup

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Merge branch 'hawkfish-window-sorting' into hawkfish-small-partition

Commit:
Issue #3954: Window Hash Groups

Switch to using adaptive partitioning.
Part 1: Start with a single ungrouped set of data.

Commit:
Issue #3954: Window Hash Groups

Switch to using adaptive partitioning.
Part 2: Use large memory footprint as a trigger for hash grouping.

Commit:
Issue #3954: Window Hash Groups

Trigger and test grouping logic.

Commit:
Merge from master (Allocator hole-cutting).

Commit:
Issue #3954: Window Hash Groups

Fix pedantic errors.

hawkfish commented:
This looks like some nasty combination of `OVER()`, external sorting and `VARCHAR` payloads, and I will have to come back to it next week.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Issue #3954: Window Hash Groups

Align row and heap pages for scanning.
Fix a number of other silly problems.

Commit:
Merge branch 'master' into hawkfish-small-partition

Commit:
Issue #3954: Window Hash Groups

Fix payload indexing typo.
Handle missing group counts.
Remove swizzling "optimisation".

hawkfish commented:
The regression failure is a CI image download problem. It passed on my fork.

Commit:
Merge branch 'master' into hawkfish-small-partition

hawkfish commented:
I think this is ready to go now.

Mytherin commented:
Thanks!
